File: src/main/java/fun/vs/fw/demo/merchantdirectory/MerchantDirectoryAdapter.java
package fun.vs.fw.demo.merchantdirectory;


import fun.vs.fw.demo.domain.CategorizedTransaction.ExpenseCategory;
import fun.vs.fw.demo.domain.MerchantDirectory;
import fun.vs.fw.demo.domain.Transaction;

public class MerchantDirectoryAdapter implements MerchantDirectory {

    private final MerchantDirectoryService merchantDirectoryService;

    public MerchantDirectoryAdapter(MerchantDirectoryService merchantDirectoryService) {
        this.merchantDirectoryService = merchantDirectoryService;
    }

    @Override
    public ExpenseCategory getFor(Transaction.MerchantCategoryCode code) {
        final var merchantInfo = merchantDirectoryService.getCategoryForMerchant(code.value());
        return new ExpenseCategory(merchantInfo.category());
    }
}

File: fvf4k/src/main/kotlin/fvf4k/demo/infra/merchantdirectory/MerchantDirectoryAdapter.kt
package fvf4k.demo.infra.merchantdirectory

import arrow.core.raise.catch
import arrow.core.raise.context.Raise
import arrow.core.raise.context.raise
import fvf4k.demo.domain.ApplicationError
import fvf4k.demo.domain.CouldNotCategorizeTransaction
import fvf4k.demo.domain.MerchantDirectory
import fvf4k.demo.domain.model.ExpenseCategory
import fvf4k.demo.domain.model.MerchantCategoryCode

class MerchantDirectoryAdapter(val merchantDirectoryService: MerchantDirectoryService) : MerchantDirectory {
    context(_: Raise<ApplicationError>)
    override fun getFor(mcc: MerchantCategoryCode): ExpenseCategory {
        val merchantInfo = catch(
            block = { merchantDirectoryService.getMerchantCategoryCode(mcc.value) },
            catch = { exception ->
                raise(CouldNotCategorizeTransaction("could not retrieve merchant info for mcc='${mcc.value}': ${exception.message}"))
            }
        )
        return ExpenseCategory.Companion(merchantInfo.category)
    }
}